### ISSUE TITLE
Suggestions for improving perf in a multi threaded application?

### ISSUE THREAD

**Issue by jaybert** [mathnet/mathnet-numerics#317]
Hi,

Hopefully this is the right place to post this, apologies if it isn't.

I have a ,net (c# and f#) application that uses mathnet numerics currently using the Intel native mkl provider and I'm running into some issues with parallelization.  The app itself is parellized and each thread is using mathnet numerics for calculations (really just vector and matrix multiplication / division).  Reading through the Intel mkl documentation, it seems like it's not really meant to be used in multi threaded applications since the mkl is itself already parallelized.   I've tried some things like limit each instance to be singlethreaded(), turned off hyper threading on the CPUs, without much luck on performance.  Looking at the resources nothing seems very taxed.

I tried doing some performance analysis on the app and there seems to be a large amount of wait time attributed to the libio* dll which I believe is just what the Intel native mkl uses for parellization.  I'm pretty new to analyzing bottlenecks/performance so it's possible I'm just reading it wrong other than I see a lot of waits on each thread pointing to that dll.  

Any suggestions on other things I can try?  For a variety of reasons not paralleling the application is not really an option.

Thanks!  Really appreciate all the hard work everyone puts into this!

**Comment by cdrnet** [comment]
Just to clarify, with "limit each instance to be singlethreaded()" you mean you've set `Control.UseSingleThread()` within Math.NET Numerics?

**Comment by jaybert** [comment]
Yes that's correct.  Called Control.UseSingleThread().  Also tried setting the env variable to tell the native mkl to run single threaded as I wasn't sure if UseSingleThread() was configured for the native MKL.  This was tested using 3.7.0 and the corresponding Intel native MKL updated at the same time
